### ISSUE TITLE
conan create doesn't use default_options in test_package

### ISSUE THREAD

**Issue by avf** [conan-io/conan#2008]
I'm trying to create a new conan package. In my package recipe `conanfile.py`, I have added an option like this:

```
    # package recipe conanfile.py
    options = {"cmake_options": "ANY"}
    default_options = "cmake_options="
```

In my consumer `conanfile.py` for the `test_package`, I'm specifying the options like this:
```
    # consumer conanfile.py in test_package
    default_options = "MyPackage:cmake_options=-DSPECIAL_FLAG=OFF -DSPECIAL_SECOND_FLAG=ON"
```
When I use `conan create` to create and test the package, the options are never passed, when I try printing them in the package recipe, they're empty. However, when I explicitly specify `MyPackage` in the `requires` attribute of the consumer `conanfile.py` and use `conan install`, everything works fine.

Is this a bug, or am I doing something wrong?

**Comment by memsharded** [comment]
Yes, there is another recommended approach. I will explain first the rationale:
- Conan create package is basically running an export, then an install with ``--build``, and then, it will go to the ``test_package`` folder, and consume the existing package. The consumption of the existing package is also a ``conan install`` to install the dependencies, then a ``conan build`` finally running the ``test()`` method
- There is almost no magic in the ``test_package`` folder test project. It is a normal consuming project of an existing package
- The first install, it is done with the default options declare in the package recipe. The ``test_package`` folder is not used at all at this point. This is why you are not seen the flags
- Then the second install, done at the ``test_package`` will try to use those flags, and indeed might require a different binary than the one that was created in the first install

So the configuration should be the same for both. You can achieve that:

- Passing the option in the command line, not in the test_package/conanfile

```bash
$ conan create user/testing -o Hello:cmake_options=value
```

- Putting the option in a profile, then using the profile. 

In both cases, you basically decouple the configuration and the logic. The logic is in the conanfiles, and the configuration is in profiles, or in the command line or project scripts. It is reasonable to have defaults for your own package, but defaulting your dependencies options might not be the most maintainable approach.

Also, I would suggest trying to better model your binaries. I know the ``ANY`` sounds useful for an option, and allows to inject like arbitrary flags, but it might have other problems. For example if you provide: ``-DSPECIAL_FLAG=OFF -DSPECIAL_SECOND_FLAG=ON`` and ``-DSPECIAL_SECOND_FLAG=ON -DSPECIAL_FLAG=OFF`` the value for the option will be different, and you will have two different binaries created. Try to model what those special flags are, what do they mean to your package, and have an option for each one of them.

Please tell me if all the above makes sense :) 
Thanks!

**Comment by avf** [comment]
Hi @memsharded, thanks for your detailed response! 
I see, I didn't know about the `create` command doing an install twice. The problem is that my package doesn't compile on all systems without specifying any options. I'm actually trying to create a conan package for the latest OpenCV (3.3.1) that also includes the `opencv_contrib` extra modules. Some of these modules don't compile for macOS, for example, so you have to disable them for that OS. However, I want my conan package to be flexible enough so that the consumer can decide which modules they want. It didn't really make sense to me to disable some arbitrary modules by default in the package recipe.

Also, thank you about your suggestion for modelling the binaries. Unfortunately, OpenCV has a ton of cmake options that I'd have to model individually and this might change in future updates, so I'd have to maintain them too. I want to keep the package recipe as simple as possible, while providing most flexibility for the consumer. This comes with the trade off you mentioned of potentially creating multiple binaries for the same configuration, but as far as I understand this is really only a matter of some additional build time and disk space. Or is there any other potential issue with this approach that I haven't thought of?

Thanks again for your help and your work on conan!

**Comment by memsharded** [comment]
Yes, that is true, I am well aware about the ``opencv_contrib`` extra modules, and how it needs to be compiled with the main repo. However, if the purpose is to build something really generic, for the larger conan community of users, it might make sense to go that extra mile and try to work to really modularize that, so the ``opencv_contrib`` modules can be distributed in their own packages, which depend on main opencv. Modeling the existence of components based on options is actually an indicator that real modularization is needed. Might need the help of opencv maintainers, I guess...

Also, copying @solvingj , just in case the bincrafters community were already working on some package for OpenCV already.

**Comment by solvingj** [comment]
Thanks @memsharded, it’s on our list, but we put it off due to its complexity, nobody on team seems to need it themselves, and it didn’t have the most requests. That said, i still consider it a high priority long-term.  I definitely invite @avf to come chat with us on slack. We’d love to collaborate with you on it. As @memsharded said, making every module a separate package is definitely the way to go. Glad we’re discussing before you got too far along on a bad path.

**Comment by avf** [comment]
@solvingj: Wow, completely modularizing OpenCV with Conan packages seems like an insane amount of work - if I understand correctly, you'd also have to create new build files, because of the way the CMakeFiles for OpenCV are currently set up. This sounds like months of work and would have to be maintained for each new OpenCV release - unfortunately out of scope for the company I work at.

However, the following recipe I've made now works pretty well, I've tested it on Linux, Windows and macOS:

```
from conans import ConanFile, CMake, tools

class OpencvConan(ConanFile):
    name = "OpenCV"
    version = "3.3.1"
    license = "http://opencv.org/license.html"
    url = "http://gitlab.internal/core/low-level-sdk"
    description = "OpenCV: Open Source Computer Vision Library"
    settings = "os", "compiler", "build_type", "arch"
    generators = "cmake"
    options = {"cmake_options": "ANY"}
    default_options = "cmake_options="
    build_policy = "missing"

    def source(self):
        self.run("git clone https://github.com/opencv/opencv.git")
        self.run("cd opencv && git checkout tags/" + self.version)
        self.run("git clone https://github.com/opencv/opencv_contrib.git")
        self.run("cd opencv_contrib && git checkout tags/" + self.version)

    def build(self):
        cmake = CMake(self)
        cmake_options = """
                        -DOPENCV_EXTRA_MODULES_PATH=%s
        """ % (self.source_folder + "/opencv_contrib/modules")
        cmake_options += str(self.options.cmake_options)
        cmake_options = str.replace(cmake_options, "\n", " ")
        # print(cmake_options)
        cmake_command = "cmake %s/opencv %s %s" % (self.source_folder, cmake.command_line, cmake_options)
        self.run(cmake_command)
        self.run("cmake --build . --target install %s" % cmake.build_config)

    def package(self):
        self.copy(pattern="*.h*", dst="include", src="include", keep_path=True)
        self.copy("*.dll", dst="bin", keep_path=False)
        self.copy("*.so", dst="lib", keep_path=False)
        self.copy("*.dylib", dst="lib", keep_path=False)
        self.copy("*.lib", dst="lib", keep_path=False)
        self.copy("*.pdb", dst="lib", keep_path=False)
        self.copy("*.a", dst="lib", keep_path=False)

    def package_info(self):
        self.cpp_info.libs = tools.collect_libs(self)
        self.cpp_info.exelinkflags.append("-framework Accelerate")
        self.cpp_info.exelinkflags.append("-framework OpenCL")
```

I've tested it with the following consumer `conanfile.py`:

```
from conans import ConanFile, CMake, tools
import os

class INSLowLevelSDK(ConanFile):
    settings = "os", "compiler", "build_type", "arch"
    generators = "cmake"
    requires = "OpenCV/3.3.1@ins/stable"
    opencv_cmake_options = """
            -DOPENCV_FP16_DISABLE=TRUE
            -DBUILD_SHARED_LIBS=OFF
            -DBUILD_PERF_TESTS=OFF
            -DBUILD_TESTS=OFF
            -DBUILD_EXAMPLES=OFF
            -DBUILD_DOCS=OFF
            -DBUILD_FAT_JAVA_LIB=OFF
            -DBUILD_opencv_apps=OFF
            -DBUILD_ANDROID_EXAMPLES=OFF
            -DBUILD_ANDROID_SERVICE=OFF
            -DBUILD_opencv_stitching=OFF
            -DBUILD_opencv_videostab=OFF
            -DBUILD_opencv_world=OFF
            -DBUILD_opencv_ccalib=OFF
            -DBUILD_opencv_cvv=OFF
            -DBUILD_opencv_contrib_world=ON
            -DBUILD_opencv_dnn=OFF
            -DBUILD_opencv_fuzzy=OFF
            -DBUILD_opencv_hdf=OFF
            -DBUILD_opencv_line_descriptor=OFF
            -DBUILD_opencv_matlab=OFF
            -DBUILD_opencv_optflow=OFF
            -DBUILD_opencv_plot=OFF
            -DBUILD_opencv_reg=OFF
            -DBUILD_opencv_rgbd=OFF
            -DBUILD_opencv_saliency=OFF
            -DBUILD_opencv_bioinspired=OFF
            -DBUILD_opencv_sfm=OFF
            -DBUILD_opencv_stereo=OFF
            -DBUILD_opencv_structured_light=OFF
            -DBUILD_opencv_surface_matching=OFF
            -DBUILD_opencv_ximgproc=OFF
            -DBUILD_opencv_xobjdetect=OFF
            -DBUILD_opencv_xphoto=OFF
            -DBUILD_opencv_aruco=ON
            -DWITH_QT=OFF
            -DWITH_QTKIT=OFF
            -DWITH_CAROTENE=OFF
            -DWITH_OPENEXR=OFF
            -DBUILD_OPENEXR=OFF
            -DWITH_IPP=OFF
            -DWITH_IPP_A=OFF
            -DWITH_JASPER=OFF
            -DBUILD_JASPER=OFF
            -DWITH_TIFF=OFF
            -DBUILD_TIFF=OFF
            -DWITH_PNG=OFF
            -DBUILD_PNG=OFF
            -DAPPLE_FRAMEWORK=ON
            -DBUILD_ZLIB=ON
            -DENABLE_NEON=ON
            -DWITH_CUDA=OFF
            -DWITH_EIGEN=OFF
            -DWITH_OPENCL=OFF
            -DBUILD_opencv_cudev=OFF
    """
    if tools.os_info.is_windows:
        opencv_cmake_options += " -DWITH_TBB=OFF "
        opencv_cmake_options += " -DBUILD_TBB=OFF "
    else:
        opencv_cmake_options += " -GNinja "
        opencv_cmake_options += " -DWITH_TBB=ON "
        opencv_cmake_options += " -DBUILD_TBB=ON "
    default_options = "OpenCV:cmake_options=" + str.replace(opencv_cmake_options, "\n", " ")

    def build(self):
        cmake = CMake(self)
        # Current dir is "test_package/build/<build_id>" and CMakeLists.txt is in "test_package"
        cmake.configure(source_dir=self.conanfile_directory, build_dir="./")
        cmake.build()

    def imports(self):
        self.copy("*.dll", dst="bin", src="bin")
        self.copy("*.dylib*", dst="bin", src="lib")
        self.copy('*.so*', dst='bin', src='lib')

    def test(self):
        os.chdir("bin")
        self.run(".%sexample" % os.sep)
```

The only downside (other than being very ugly) I can think of with this approach is that someone might have to rebuild the same package because they specified the same CMake options in a different order.

This is getting pretty off topic now though, so maybe we should close and/or rename this issue.

**Comment by memsharded** [comment]
@avf Yes, modularizing OpenCV is offtopic, but I wanted at least to kick off the conversation about it. You can follow the efforts about it in the bincrafters community (slack, twitter, github, bintray)

Regarding your original question, I understand your trade-offs, and I am good with them, if they fit to your purpose. For the general public or for conan-center, it is not a solid approach, because the potential explosion in binaries is too high. But if it is good for you, why not?

I would just suggest that you moved the default options settings from the ``test_package/conanfile.py`` to the main ``conanfile.py``. That way it should work without issues. And it is very important that they are there, because that is the way to indicate consumers which options in cmake they can change. If users type:

```bast
$ conan get OpenCV/3.1.1@user/channel  #either in local or in -r=remote
```
What they get listed is the main conanfile.py. Reading it, they can deduce easily how to change some flags. Otherwise, they probably need to search and navigate to the repo to have a look to ``test_package/conanfile.py``.

**Comment by avf** [comment]
Sure, sorry, I didn't mean to sound dismissive, I just didn't want to hijack this issue with a very specific problem. I think it'd be awesome to fully modularize OpenCV with conan, that thing is a beast and it'd be great to tame it!

Yeah, I agree, for conan-center my approach wouldn't be great. This is more something we'll use internally at our company.

Thanks for your suggestion about putting the default options in the package recipe. I just thought it makes more sense to keep the package recipe as minimal as possible and let the consumer define everything - it's really the consumers responsibility which modules they want to include.

Another question: Is there a way to change the `default_options` dynamically in the consumer `conanfile.py` depending on the settings? For example, when building for Windows, I have to specify a different setting (see above). I've tried using the `configure()` method, but it doesn't seem to work in the consumer conanfile. I use a workaround above with `tools.os_info.is_windows`, but this will break once I'm cross-compiling.

**Comment by memsharded** [comment]
No problem, you didn't :) :) :) Fully understand what you meant.

IMHO recipes should be minimal but as self-contained as possible. In that sense, it is very good that they contain some sensible defaults for the options it is using, so actually releases the consumers to figure out. A large recipe is not a big problem, but not being evident which options you can use, is more burden to consumers than the other approach.

No, you cannot change the default_options in the consumer, but you can directly set options (like ``self.options.myoption=myvalue`` if you want in the ``configure()``. The effect should be the same.

**Comment by memsharded** [comment]
While removing ``conan test_package`` we have hit this issue while fixing broken tests. Indeed, if we want to support some use cases like having "build_requires" that have to be built in the ``test_package`` conanfiles, we need to revert this behavior. So then, everything in test_package will be applied, and the main use case that originated your first comment, will work as you expected: default_options defined in test_package will also be used.

I have proposed it in #2081, that doesn't mean that it will be done, but you might want to follow it. Thanks!

**Comment by memsharded** [comment]
This should be solved now, will be released in 0.30

**Comment by memsharded** [comment]
Released in 0.30! Please @avf have a try, now options in ``test_package/conanfile.py`` should be properly applied as you wanted. Comment or reopen if necessary, thanks!
